src/runtime/device_interface.cpp

```diff
@@ -387,4 +390,176 @@ WEAK void halide_device_and_host_free_as_destructor(void *user_context, void *ob
 WEAK void halide_device_host_nop_free(void *user_context, void *obj) {
 }
 
+WEAK int halide_default_buffer_copy(void *user_context, struct halide_buffer_t *src,
+                                    const struct halide_device_interface_t *dst_device_interface,
+                                    struct halide_buffer_t *dst) {
+
+    const bool from_device = src->device && (src->device_dirty() || !src->host);
+    const bool to_device = dst_device_interface;
+    const bool from_host = !from_device;
+    const bool to_host = !to_device;
+
+    debug(user_context)
+        << "halide_default_buffer_copy\n"
+        << " source: " << *src << "\n"
+        << " dst: " << *dst << "\n";
+
+    // We consider four cases, and decompose each into simpler cases
+    // using intermediate host memory in the src or dst buffers.
+    int err = 0;
+    if (from_device && to_device) {
+        // dev -> dev via dst host memory
+        debug(user_context) << " dev -> src via src host memory\n";
+        err = src->device_interface->impl->buffer_copy(user_context, src, NULL, dst);
+        err = err || copy_to_device_already_locked(user_context, dst, dst_device_interface);
+    } else if (from_device && to_host) {
+        // dev -> host via src host memory
+        debug(user_context) << " dev -> host via src host memory\n";
+        err = copy_to_host_already_locked(user_context, src);
+        err = err || halide_default_buffer_copy(user_context, src, NULL, dst);
+    } else if (from_host && to_device) {
+        // host -> dev via dst host memory
+        debug(user_context) << " host -> dev via dst host memory\n";
+        err = halide_default_buffer_copy(user_context, src, NULL, dst);
+        err = err || copy_to_device_already_locked(user_context, dst, dst_device_interface);
+    } else {
+        // host -> host
+        debug(user_context) << " host -> host\n";
+        device_copy copy = make_buffer_copy(src, true, dst, true);
+        copy_memory(copy, user_context);
+        dst->set_host_dirty();
+    }
+
+    return err;
+}
+
+WEAK int halide_buffer_copy(void *user_context, struct halide_buffer_t *src,
+                            const struct halide_device_interface_t *dst_device_interface,
+                            struct halide_buffer_t *dst) {
+    ScopedMutexLock lock(&device_copy_mutex);
+
+    debug(user_context) << "halide_buffer_copy:\n"
+                        << " src " << src << "\n"
+                        << " interface " << dst_device_interface << "\n"
+                        << " dst " << dst << "\n";
+
+    const struct halide_device_interface_t *src_device_interface = src->device_interface;
+
+    int err = 0;
+
+    // Fix up the dst device interface
+    if (dst_device_interface &&
+        dst_device_interface != dst->device_interface) {
+        err = halide_device_free(user_context, dst);
+        err = err || halide_device_malloc(user_context, dst, dst_device_interface);
```

Comment:
Is it really fixed? err || x will be true or false, losing the error code stored in err.